sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler_opencl.asciidoc

```diff
@@ -240,52 +373,44 @@ int main() {
 }
 ```
 
+=== Querying supported features and extensions
 
-== Issues
-
-* How should we expose the difference between OpenCL C versions?
-  It seems like there are two aspects to the problem.
-  Applications need some way to query which versions the backend (or device)
-  supports.
-  Applications also need some way to tell the runtime which version the kernel
-  is written in.
-+
---
-One option is to define separate enumerators in `source_language` for each
-version like this:
+This example demonstrates how to query the version of OpenCL C that is
+supported, how to query the supported features, and how to query the
+supported extensions.
 
 ```
-enum class source_language : /*unspecified*/ {
-  opencl_1_0,
-  opencl_1_1,
-  opencl_2_0,
-  opencl_3_0,
-};
+#include <iostream>
+#include <sycl/sycl.hpp>
+namespace syclex = sycl::ext::oneapi::experimental;
+
+int main() {
+  sycl::queue q;
+  sycl::device q.get_device();
+
+  if (d.ext_oneapi_can_compile(syclex::source_language::opencl) {
+    std::cout << "Device supports online compilation of OpenCL C kernels\n";
+  }
+  if (d.ext_oneapi_supports_cl_c_version(syclex::opencl_c_3_0) {
+    std::cout << "Device supports online compilation with OpenCL C 3.0\n";
+  }
+  if (d.ext_oneapi_supports_cl_c_feature("__opencl_c_fp64") {
+    std::cout << "Device supports online compilation with 64-bit FP in OpenCL C\n";
+  }
+  syclex::cl_version version;
+  if (d.ext_oneapi_supports_cl_extension("cl_intel_bfloat16_conversions", &version)) {
```

Comment:
Perhaps it's an implementation detail but I'd like to mention it anyway. This function is expected to be called multiple times, possibly for each potentially supported extension. With that, it could be very expensive to use ocloc every time the function is being called. It would probably make sense to consider caching the result of the ocloc query to avoid undesirable overhead.

Reply:
Yes, I agree.  The implementation should use caching to avoid multiple redundant calls to ocloc that query the set of supported extensions.